File: R/imf/imf_function.R
# Función mínima para descargar datos del FMI (SDMX 3.0 → SDMX-CSV 1.0)
# Inspirada en eurostat_function.R, usando el flujo de imf_min.R
#
# Inputs obligatorios:
# - dataset_identifier: identificador del dataset (p. ej., "QNEA")
# - data_selection: clave SDMX (orden y códigos según el dataset), p. ej.,
#   "ESP.B1GQ.Q.SA.XDC.Q" o con múltiples países "ESP+FRA.B1GQ.Q.SA.XDC.Q".
#
# Inputs opcionales:
# - agency_identifier: p. ej., "IMF.STA"
# - dataset_version: p. ej., "+" (última versión)
# - filters: lista de filtros SDMX como c[TIME_PERIOD] = "ge:2020-Q1+le:2020-Q4".
#            Páselos como list(TIME_PERIOD = c("ge:2020-Q1","le:2020-Q4"))
#            Se convertirán a query params tipo c[TIME_PERIOD]="ge:...+le:...".
# - accept_csv_version: versión del formato SDMX-CSV (por defecto "1.0.0")
#
# Output:
# - data.frame leído de la respuesta CSV

imf_api_function <- function(
  dataset_identifier,
  data_selection,
  filters = list(),
  agency_identifier = "IMF.STA",
  dataset_version = "+",
  accept_csv_version = "1.0.0"
) {
  base_url <- "https://api.imf.org/external/sdmx/3.0/data/dataflow"

  # Construir ruta completa: {base}/{agency}/{dataset}/{version}/{key}
  # Codificamos cuidadosamente la versión ('+' -> %2B) y la clave
  url <- paste0(
    base_url, "/",
    agency_identifier, "/",
    dataset_identifier, "/",
    utils::URLencode(dataset_version, reserved = TRUE), "/",
    utils::URLencode(data_selection, reserved = TRUE)
  )

  # Construir filtros como c[DIM]=v1+v2 (TIME_PERIOD ge/le, etc.)
  if (!is.list(filters)) stop("'filters' debe ser una lista nombrada de dimensiones -> valores")
  filter_params <- list()
  if (length(filters)) {
    for (nm in names(filters)) {
      vals <- filters[[nm]]
      if (is.null(vals) || length(vals) == 0) next
      # Join con '+' (IMF acepta ge:...+le:... y múltiples valores con '+')
      filter_params[[paste0("c[", nm, "]")]] <- paste(vals, collapse = "+")
    }
  }

  headers <- httr::add_headers(
    `Cache-Control` = "no-cache",
    Accept = paste0("application/vnd.sdmx.data+csv;version=", accept_csv_version)
  )

  res <- httr::GET(url, query = filter_params, headers, httr::config(ssl_verifypeer = 0L, ssl_verifyhost = 0L))
  if (httr::http_error(res)) stop(sprintf("IMF request failed [%s]", httr::status_code(res)))

  txt <- httr::content(res, as = "text", encoding = "UTF-8")
  df <- utils::read.csv(text = txt, stringsAsFactors = FALSE)
  return(df)
}

File: python/imf/imf_function.py
from __future__ import annotations

import io
from typing import Dict, Iterable

import pandas as pd
import requests


def imf_api_function(
    dataset_identifier: str,
    data_selection: str,
    filters: Dict[str, Iterable[str]] | None = None,
    agency_identifier: str = "IMF.STA",
    dataset_version: str = "+",
    accept_csv_version: str = "1.0.0",
    base_url: str = "https://api.imf.org/external/sdmx/3.0/data/dataflow",
) -> pd.DataFrame:
    """
    Descargar datos desde el FMI (SDMX 3.0 CSV) y devolver un DataFrame.
    - dataset_identifier: p. ej., "QNEA"
    - data_selection: clave SDMX (p. ej., "ESP.B1GQ.Q.SA.XDC.Q" o "ESP+FRA.B1GQ.Q.SA.XDC.Q")
    - filters: dict como { 'TIME_PERIOD': ['ge:2020-Q1','le:2020-Q4'] }
    """
    if filters is None:
        filters = {}

    # Construir la URL SDMX 3.0 CSV: base/agency/dataset/version/key
    # Codificar conservando reservados útiles (+,*,:,/ ,)
    from urllib.parse import quote

    def urlencode_keep_reserved(s: str) -> str:
        return quote(str(s), safe="+*:,/,")

    url = (
        f"{base_url}/"
        f"{agency_identifier}/"
        f"{dataset_identifier}/"
        f"{urlencode_keep_reserved(dataset_version)}/"
        f"{urlencode_keep_reserved(data_selection)}"
    )

    # Filtros como c[DIM] = v1+v2 (unir con '+')
    params: Dict[str, str] = {}
    for dim, values in filters.items():
        if values is None:
            continue
        if isinstance(values, (list, tuple, set)):
            joined = "+".join(map(str, values))
        else:
            joined = str(values)
        params[f"c[{dim}]"] = joined

    headers = {
        "Accept": f"application/vnd.sdmx.data+csv;version={accept_csv_version}",
        "Cache-Control": "no-cache",
    }

    resp = requests.get(url, params=params, headers=headers, timeout=180)
    resp.raise_for_status()

    text_stream = io.StringIO(resp.content.decode("utf-8"))
    df = pd.read_csv(text_stream)
    return df
